Update sorbet-typed to remove directories and make sorbet-typed repo configurable

connorshea opened the pull request:
This includes two changes, I can split them into separate PRs if we want that, but the first is pretty minor so I figured it was fine to include here:

- Make it possible to pull in sorbet-typed RBIs from a configurable sorbet-typed repo. This can be done by setting the `SRB_SORBET_TYPED_REPO` environment variable to a git URL. This is useful for testing major changes to sorbet-typed, where previously we had to go into `~/.cache/sorbet/sorbet-typed/` and mess with the git repo in there.
- Make sorbet-typed delete RBI files that are no longer relevant, e.g. because the gem has been removed from the project, the RBI's version is no longer compatible with your current version of the gem, or the RBI was removed from the sorbet-typed repo entirely.

### Motivation
Fixes #1260.

I noticed this when `gem.rbi` was deleted from sorbet-typed last week, and running `srb rbi sorbet-typed` didn't delete the file. It's also relevant with Rails 6 being out, since anyone upgrading from 5.2 would have a lingering RBI for `~>5.2.0` which isn't compatible with their new Rails version.

### Test plan
There aren't any existing tests, so I figured it was fine not to add any for this command for now. You can test this by:

- Set up this branch on a project with sorbet already instantiated
- Create a file like `sorbet/rbi/sorbet-typed/lib/fake-gem/all/fake-gem.rbi`, it doesn't need to have anything in it.
- Run `srb rbi sorbet-typed`
- Note that the `fake-gem.rbi` has been deleted.

You can also verify this by adding an incompatible version to an existing gem you already have type definitions for, e.g. add `sorbet/rbi/sorbet-typed/lib/activerecord/~>9.0.0/activerecord.rbi` and run `srb rbi sorbet-typed` to delete it.

I'm sure there are ways to make this code simpler, I'd be happy to hear any feedback :)

jez commented on gems/sorbet/lib/fetch-rbis.rb:
```diff
@@ -14,12 +14,12 @@ class Sorbet::Private::FetchRBIs
 
-  SORBET_TYPED_REPO = 'https://github.com/sorbet/sorbet-typed.git'
+  SORBET_TYPED_REPO = ENV['SRB_SORBET_TYPED_REPO'] || 'https://github.com/sorbet/sorbet-typed.git'
   SORBET_TYPED_REVISION = ENV['SRB_SORBET_TYPED_REVISION'] || 'origin/master'
```
Our original intention of this `SRB_SORBET_TYPED_REVISION` environment variable was to do exactly what it seems you're trying to do in this PR. e.g.,

- cd into the cache dir
- git remote add <the fork>
- export SRB_SORBET_TYPED_REVISION="<the PR branch name>"

In general, we'd prefer to not have more than one way to do things, and

jez commented on gems/sorbet/lib/fetch-rbis.rb:
```diff
@@ -14,12 +14,12 @@ class Sorbet::Private::FetchRBIs
 
-  SORBET_TYPED_REPO = 'https://github.com/sorbet/sorbet-typed.git'
+  SORBET_TYPED_REPO = ENV['SRB_SORBET_TYPED_REPO'] || 'https://github.com/sorbet/sorbet-typed.git'
   SORBET_TYPED_REVISION = ENV['SRB_SORBET_TYPED_REVISION'] || 'origin/master'
```
Our original intention of this `SRB_SORBET_TYPED_REVISION` environment variable was to do exactly what it seems you're trying to do in this PR. e.g.,

- cd into the cache dir
- git remote add <the fork>
- export SRB_SORBET_TYPED_REVISION="<the PR branch name>"

In general, we'd prefer to not have more than one way to do things. Given this:

> where previously we had to go into ~/.cache/sorbet/sorbet-typed/ and mess with the git repo in there

it seems to me like mucking with an environment variable is already "messing with `srb` internals," so going into `~/.cache` seems like it shouldn't be a huge step. And it should always be safe to just nuke that directory if it gets into a weird git state.

connorshea replied:
I disagree that mucking with the git repo in `~/.cache/` is as easy as setting an environment variable, e.g. it's the difference between being able to just run `SRB_SORBET_TYPED_REPO=https://github.com/connorshea/sorbet-typed.git bundle exec srb rbi sorbet-typed` vs. needing to:
- `cd ~/.cache/sorbet/sorbet-typed/`
- `git remote add fork https://github.com/connorshea/sorbet-typed.git`
- `SRB_SORBET_TYPED_REVISION=fork/master bundle exec srb rbi sorbet-typed`

It does become essentially equal after the first initial setup, but I'd still prefer to be able to just use `SRB_SORBET_TYPED_REPO=https://github.com/connorshea/sorbet-typed.git bundle exec srb rbi sorbet-typed` whenever I want to pull from a specific remote.

Admittedly, I didn't know about the REVISION environment variable, and didn't really think of being able to add a git repo as a remote rather than overriding the `origin`.

I guess I'd be fine with adding the environment variable for specifying the remote, though I still think this is a simpler solution in the end.

Also, since the code below does `git fetch origin` before `git checkout -q SORBET_TYPED_REVISION`, I'm not actually sure if the checkout would work for a remote other than `origin` (unless you went into the cache directory and fetched). 🤔 I'd need to test that.

jez commented on gems/sorbet/lib/fetch-rbis.rb:
```diff
@@ -114,11 +128,12 @@ def self.main
 
     if vendor_paths.length > 0
       vendor_rbis_within_paths(vendor_paths)
+      remove_extra_rbis(vendor_paths)
```
This one doesn't make sense to me. Can we not just remove the entire folder before re-generating, rather than having to selectively delete things after the fact? It looks like we already do this for generating `sorbet/rbi/gems` and `sorbet/rbi/hidden-definitions`:

https://github.com/sorbet/sorbet/blob/c404b8915d19147da3ef68de4fe7fde2e4caf48d/gems/sorbet/lib/gem-generator-tracepoint.rb#L42

connorshea replied:
ah, yeah that would be much simpler. I'll change it to do that.

connorshea commented:
@jez I've removed the configurable sorbet-typed git repo from this PR and simplified the sorbet-typed command to just delete the directory entirely before repopulating it.
